File: ToDoApi/ToDoApp/src/app/services/todo.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface ToDoItem {
  itemID: number;
  title: string;
  completed: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class TodoService {
  private apiUrl = 'https://localhost:5117/api/todos';

  constructor(private http: HttpClient) { }

  getItems(): Observable<ToDoItem[]> {
    return this.http.get<ToDoItem[]>(this.apiUrl);
  }

  addItem(item: ToDoItem): Observable<ToDoItem> {
    return this.http.post<ToDoItem>(this.apiUrl, item);
  }

  updateItem(item: ToDoItem): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${item.itemID}`, item);
  }

  deleteItem(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: ToDoApi/ToDoApi/Services/ToDoService.cs
using ToDoApi.Models;

namespace ToDoApi.Services {
    public class ToDoService : IToDoService {
        private readonly List<ToDoItem> _Items = [];
        private int _NextID = 1;

        public List<ToDoItem> GetItems() => _Items;

        public ToDoItem GetItem(int itemID) => _Items.FirstOrDefault(i => i.ItemID == itemID);

        public ToDoItem AddItem(ToDoItem item) {
            item.ItemID = _NextID++;
            _Items.Add(item);
            return item;
        }

        public ToDoItem UpdateItem(ToDoItem item) {
            ToDoItem existingItem = GetItem(item.ItemID);
            if (existingItem != null) {
                existingItem.Title = item.Title;
                existingItem.Completed = item.Completed;
            }
            return existingItem;
        }

        public bool DeleteItem(int id) {
            ToDoItem existingItem = GetItem(id);
            return existingItem != null && _Items.Remove(existingItem);
        }
    }
}
